lib/harvester_utils/downloader.rb

```diff
@@ -18,13 +20,32 @@ def self.url_exists?(url)
     end
 
     def self.download_file(url)
-      response = Faraday.get(url)
+      conn = Faraday.new(request: { timeout: 300 })
+      response = conn.get(url)
+
       raise DownloadError, "Unexpected response status: #{response.status} for url: #{url}" if response.status != 200
-      file = Tempfile.new(rand(1_000_000).to_s)
-      file.binmode
-      file.write(response.body)
-      file.close
-      file
+
+      zip_file = nil
+      Tempfile.open do |file|
+        file.binmode
+        file.write(response.body)
+        zip_file = file
+      end
+
+      self.extract_csv(zip_file)
+    end
+
+    def self.extract_csv(zip_file)
+      Zip::File.open(zip_file) do |zip|
+        csv_file = zip.glob('*.csv')
+        raise NoCSVError, "No CSV file found in the provided zip_file: #{zip_file}" if csv_file.nil?
+
+        Tempfile.open do |file|
+          file.binmode
+          file.write(csv_file.first.get_input_stream.read)
+          file
```

Comment:
Toto je asi nebezpecne, ak ten extract ocakavame ze vrati subor tak nevieme ci to ten blok ked vrati ten subor nezavrie.